content/posts/networkx/vf2pp/ISO-feasibility-candidates/index.md

```diff
@@ -0,0 +1,156 @@
+---
+title: "ISO Feasibility & Candidates"
+date: 2022-07-11
+draft: false
+description: "
+Information about my progress on two important features of the algorithm.
+"
+tags: ["gsoc", "networkx"]
+displayInList: true
+author: ["Konstantinos Petridis"]
+
+---
+
+The previous post can be found [here]({{< relref "../node-ordering-Ti-updating" >}}), be sure to check it out, so you
+can
+follow the process step by step. Since then, another two very significant features of the algorithm have been
+implemented and tested.
+
+## Introduction
+
+As it was previously mentioned, in the ISO problem we are basically trying to create a **mapping** such that, every node
+from the first graph is matched to a node from the second graph. This searching for "feasible pairs" can be visualized
+by a tree, where each node is the candidate pair that we should examine. This can become much clearer if we take a look
+at the below figure.
+
+<center><img src="dfs.png" alt="DFS VF2++ example$"/></center>
+
+In order to check if the graphs $G_1$, $G_2$ are isomorphic, we check every candidate pair of nodes and if it is
+feasible, we extend the mapping and go deeper into the tree of pairs. If it's not feasible, we climb up and follow a
+different branch, until the mapping is extended to its full. In our example, we start by examining node 0 from G1, with
+node 0 of G2. After some checks that are skipped for simplicity (but can be perceived by observation), we decide that the
+nodes 0 and 0 are matching, so we go deeper to map the remaining nodes. The next pair is 1-1, which fails the
+feasibility check, so we have to examine a different branch as shown. The new branch is 1-2, which is feasible, so we
+continue on using the same logic until all the nodes are mapped.
+
+## Candidate Pair Selection
+
+Although in our example we use a random candidate pair of nodes, in the actual implementation we are able to target
+specific pairs that are more likely to be matched, hence boost the performance of the algorithm. The idea is that, in
+every step of the algorithm, **given a candidate**
+
+$$u\in V_1$$
+
+**we compute the candidates**
+
+$$v\in V_2$$
+
+where $V_1$ and $V_2$ are the nodes of $G_1$ and $G_2$ respectively. Now this is a puzzle that does not require a lot of
+specific knowledge on graphs or the algorithm itself. Keep up with me, and you will realize it yourself. First, let $M$
+be the mapping so far, which includes all the "covered nodes" until this point. There are actually **three** different
+types of $u$ nodes that we might encounter.
+
+### Case 1
+
+Node $u$ has no neighbors (degree of $u$ equals to zero). It can directly be realized that it would be redundant to test
```

Comment:
More (unimportant) wording suggestions, feel free to ignore!
```suggestion
Node $u$ has no neighbors (degree of $u$ equals to zero). It would be redundant to test
```